src/shogun/machine/GeneralizedLinearMachine.cpp

```diff
@@ -0,0 +1,299 @@
+	else
+	{
+		w = SGVector<float64_t>(n_features);
+		for (int i = 0; i < n_features; i++)
+			w = 1 / (n_features + 1) * distribution(generator);
+	}
+	
+	//Iterative Updates
+	int n_iter = 0;
+	for (int i = 0; i < m_max_iter; i++)
+	{
+		n_iter++;
+		SGVector<float64_t> w_old(m_w);
+		SGVector<float64_t> gradient_w = compute_grad_L2_loss_w(X, y, w, bias);
+		float64_t gradient_bias = compute_grad_L2_loss_bias(X, y, w, bias);
+
+		//Update
+        w = linalg::add(w, gradient_w, 1.0, -1*m_learning_rate);
+		if(m_fit_intercept)
+			bias = bias - m_learning_rate * gradient_bias;
+	
+		//Apply proximal operator
+		w = apply_proximal_operator(w, m_lambda * m_alpha);
+
+		//Convergence by relative parameter change tolerance
+		float64_t norm_update = linalg::norm(linalg::add(w, w_old, 1.0, -1.0));
+		if(i > 0 && (norm_update / linalg::norm(w)) < m_tolerance)
+			break;		
+	}
+
+	//Update the estimated variables
+	m_w = w;
+	m_bias = bias;
+
+	return true;
+}
+
+//TODO
+std::shared_ptr<RegressionLabels> GeneralizedLinearMachine::apply_regression(std::shared_ptr<Features> data)
+{
+	
+}
+
+//TODO
+SGVector<float64_t> GeneralizedLinearMachine::apply_get_outputs(std::shared_ptr<Features> data)
+{
+
+}
+
+//TODO
+bool GeneralizedLinearMachine::train_machine(std::shared_ptr<const DenseFeatures<float64_t>> feats)
+{
+
+}
+
+SGVector<float64_t> GeneralizedLinearMachine::get_w() const
+{
+	return m_w;
+}
+
+void GeneralizedLinearMachine::set_w(const SGVector<float64_t> w)
+{
+	m_w = w;
+}
+
+void GeneralizedLinearMachine::set_bias(float64_t b)
+{
+	m_bias=b;
+}
+
+float64_t GeneralizedLinearMachine::get_bias() const
+{
+ 	return m_bias;
+}
+
+void GeneralizedLinearMachine::set_features(std::shared_ptr<DotFeatures> feat)
+{
+ 	features=std::move(feat);
+}
+
+std::shared_ptr<DotFeatures> GeneralizedLinearMachine::get_features()
+{
+	return features;
+}
+
+SGVector<float64_t> GeneralizedLinearMachine::conditional_intensity(SGMatrix<float64_t> X, SGVector<float64_t> w, float64_t bias)
+{
+	SGVector<float64_t> z = compute_z(X, w, bias);
+	SGVector<float64_t> result = non_linearity(z);
+	return result;
+}
+
+SGVector<float64_t> GeneralizedLinearMachine::compute_z(SGMatrix<float64_t> X, SGVector<float64_t> w, float64_t bias)
+{
+	SGVector<float64_t> z = linalg::matrix_prod(X, w, false);
+	return z;
+}
+
+SGVector<float64_t> GeneralizedLinearMachine::non_linearity(SGVector<float64_t> z)
+{
+	SGVector<float64_t> result;
+	switch (distribution)
+	{
+	case POISSON:
+		result = SGVector<float64_t>(z);
+		float64_t bias = 0;
+
+		if(m_fit_intercept)
+			bias = (1 - m_eta) * exp(m_eta);
+
+		for (int i = 0; i < z.vlen; i++)
+		{
+			if(z[i]>m_eta)
+				result[i] = z[i] * exp(m_eta) + bias;
+			else
+				result[i] = exp(z[i]);
+		}
+		break;
+	
+	default:
+		break;
+	}
+	return result;
+}
+
+SGVector<float64_t> GeneralizedLinearMachine::gradient_non_linearity(SGVector<float64_t> z)
+{
+	SGVector<float64_t> result;
+	switch (distribution)
+	{
+	case POISSON:
```

Comment:
I think since nonlinearity and its gradient have to match, why not implement them in a (potentially local) class. Then it also can be unit tested more easily

Reply:
Okay. By local class you mean a class inside a function right?

Where should I implement that? I'm slightly confused.

I should make separate classes for Poisson, Gamma, etc right? And implement `non_linearity()`, `gradient_non_linearity()` etc in those classes right?

Reply:
you could do that in classes or functions, but the point is to *not* do it inside the training method, because that creates hard to read code. Keep it modular.